test/integration/scheduler_perf/scheduler_perf_test.go

```diff
@@ -51,258 +53,475 @@ var (
 	}
 )
 
-// testCase configures a test case to run the scheduler performance test. Users should be able to
-// provide this via a YAML file.
-//
-// It specifies nodes and pods in the cluster before running the test. It also specifies the pods to
-// schedule during the test. The config can be as simple as just specify number of nodes/pods, where
-// default spec will be applied. It also allows the user to specify a pod spec template for more
-// complicated test cases.
-//
-// It also specifies the metrics to be collected after the test. If nothing is specified, default metrics
-// such as scheduling throughput and latencies will be collected.
+// testCase defines a set of test cases that intend to test the performance of
+// similar workloads of varying sizes with shared overall settings such as
+// feature gates and metrics collected.
 type testCase struct {
-	// description of the test case
-	Desc string
-	// configures nodes in the cluster
-	Nodes nodeCase
-	// configures pods in the cluster before running the tests
-	InitPods []podCase
-	// configures the test to now wait for init pods to schedule before creating
-	// test pods.
-	SkipWaitUntilInitPodsScheduled bool
-	// pods to be scheduled during the test.
-	PodsToSchedule podCase
-	// optional, feature gates to set before running the test
+	// Name of the testCase.
+	Name string
+	// Feature gates to set before running the test. Optional.
 	FeatureGates map[featuregate.Feature]bool
-	// optional, replaces default defaultMetricsCollectorConfig if supplied.
+	// List of metrics to collect. Optional, defaults to
+	// defaultMetricsCollectorConfig if unspecified.
 	MetricsCollectorConfig *metricsCollectorConfig
+	// List of workloads to run under this testCase.
+	Workloads []*workload
+	// TODO(adtac): reduce config toil by having a default pod and node spec per
+	// testCase? CreatePods and CreateNodes ops will inherit these unless
+	// manually overridden.
 }
 
-type nodeCase struct {
-	Num              int
+// workload is a subtest under a testCase that tests the scheduler performance
+// for a certain ordering of ops. The set of nodes created and pods scheduled
+// in a workload may be heterogenous.
+type workload struct {
+	// Name of the workload.
+	Name string
+	// A list of ops to be executed serially. Each element of the list must be
+	// createNodesOp, createPodsOp, or barrierOp.
+	Ops []op
+}
+
+func (w *workload) collectsMetrics() bool {
+	for _, op := range w.Ops {
+		if op.realOp.collectsMetrics() {
+			return true
+		}
+	}
+	return false
+}
+
+// op is a dummy struct which stores the real op in itself.
+type op struct {
+	realOp realOp
+}
+
+// UnmarshalJSON is a custom unmarshaler for the op struct since we don't know
+// which op we're decoding at runtime.
+func (op *op) UnmarshalJSON(b []byte) error {
+	possibleOps := []realOp{
+		&createNodesOp{},
+		&createPodsOp{},
+		&barrierOp{},
+		// TODO(adtac): add a sleep timer op to simulate user action?
```

Comment:
ditto, perhaps open a single issue listing all possible improvements.

Reply:
https://github.com/kubernetes/kubernetes/issues/93793

I think having separate issues might be easier from a discussion point-of-view. I don't have a strong case for the timer op for example, so we can close that issue if everyone thinks it'd be pointless without affecting the other issues.